veloren: init at 0.16.0

rnhmjoj opened the pull request:
## Description of changes

An actual package of veloren, not just the downloader/launcher (see https://github.com/NixOS/nixpkgs/issues/95226).
This is built from source using the stable rust toolchain, against all odds.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

rnhmjoj committed:
veloren: init at 0.16.0

rnhmjoj commented on pkgs/by-name/ve/veloren/package.nix:
```diff
@@ -0,0 +1,125 @@
+  };
+
+  # Game assets, should not to be cached by hydra
+  assets = lib.dontDistribute (runCommand "veloren-assets" {} ''
```
Note: I'm [told](https://matrix.to/#/%23dev%3Anixos.org/%24zw9una3U1xAgRtK-kmsdhMKb4deACquUjELl2nxV0qY?via=maxwell.ydns.eu&via=matrix.org&via=snug.si&via=turbo.ooo) this is probably not going to work.
Anyway, the assets are just 321MB, which I suppose is acceptable for the cache.

tomodachi94 commented on pkgs/by-name/ve/veloren/package.nix:
```diff
@@ -0,0 +1,127 @@
+  pname = "veloren";
+  inherit src version;
+
+  cargoLock.lockFile = ./Cargo.lock;
```
Is there a reason we can't use the upstream lockfile? Using it might make this package easier to maintain, as well as [avoiding the undesired side effects listed here](https://github.com/NixOS/nixpkgs/issues/327064).

```nix
cargoLock.lockFile = src + "/Cargo.lock";
```

rnhmjoj replied:
No idea, probably I just didn't think of this.

rnhmjoj replied:
It seems to work.

Aleksanaa replied:
This is input from derivation, and isn't allowed on ofborg and hydra.

rnhmjoj replied:
Here's what the [manual](https://nixos.org/manual/nixpkgs/stable/#compiling-rust-applications-with-cargo) says:

> Exception: If the application has cargo git dependencies, the cargoHash/cargoSha256 approach will not work, and you will need to copy the Cargo.lock file of the application to nixpkgs and continue with the next section for specifying the options of the cargoLock section.

So, that's why I had to include the lock file.

tomodachi94 commented on pkgs/by-name/ve/veloren/package.nix:
```diff
@@ -0,0 +1,127 @@
+
+  postPatch = ''
+    # Patch cargo dependencies
+    patch "$cargoDepsCopy/vek-0.15.8/build.rs" '${./fix-vek-on-stable.patch}'
```
Maybe this dependency patch should be included in [pkgs/build-support/rust/default-crate-overrides.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/rust/default-crate-overrides.nix), so that other packages can benefit from this?

rnhmjoj replied:
Nah, it's a very ugly hack: it's forcing vek to build in unstable mode, which wouldn't make sense on its own. Why? Because even if we're in rust stable we had set RUSTC_BOOTSTRAP=1 (which is an even bigger hack and highly discouraged) to enable some unstable feature that get us barely far enough to compile veloren.

tomodachi94 commented on pkgs/by-name/ve/veloren/package.nix:
```diff
@@ -0,0 +1,127 @@
+  # Some tests require internet access
+  doCheck = false;
+
+  postFixup = ''
+    # Add required but not explicitly requested libraries
+    patchelf --add-rpath '${lib.makeLibraryPath [
+      xorg.libX11
+      xorg.libXi
+      xorg.libXcursor
+      xorg.libXrandr
+      vulkan-loader
+    ]}' "$out/bin/veloren-voxygen"
+  '';
```
Is there a reason we don't use `autoPatchelfHook` here?

rnhmjoj replied:
There's just not much to gain: I'd still have to specify the dependencies and it would also patch the other binary, which is unnecessary.

It would be useful if it could fail on a missing library, but I don't think it can detect them due to the way these are loaded at runtime depending on the platform.

rnhmjoj replied:
It doesn't seem to work well:
```
       > auto-patchelf: 4 dependencies could not be satisfied
       > error: auto-patchelf could not satisfy dependency libgcc_s.so.1 wanted by /nix/store/175afjmxpg943gil13s23h0mksrzfsgr-veloren-0.15.0/bin/veloren-server-cli
       > error: auto-patchelf could not satisfy dependency libshaderc_shared.so.1 wanted by /nix/store/175afjmxpg943gil13s23h0mksrzfsgr-veloren-0.15.0/bin/veloren-voxygen
       > error: auto-patchelf could not satisfy dependency libstdc++.so.6 wanted by /nix/store/175afjmxpg943gil13s23h0mksrzfsgr-veloren-0.15.0/bin/veloren-voxygen
       > error: auto-patchelf could not satisfy dependency libgcc_s.so.1 wanted by /nix/store/175afjmxpg943gil13s23h0mksrzfsgr-veloren-0.15.0/bin/veloren-voxygen
```
I did 
```patch
--- a/pkgs/by-name/ve/veloren/package.nix
+++ b/pkgs/by-name/ve/veloren/package.nix
@@ -67,7 +68,11 @@ rustPlatform.buildRustPackage {
     substituteInPlace common/assets/src/lib.rs --subst-var-by assets '${assets}'
   '';
 
-  nativeBuildInputs = [ pkg-config ];
+  nativeBuildInputs = [
+    pkg-config
+    autoPatchelfHook
+  ];
+
   buildInputs = [
     alsa-lib
     udev
@@ -75,6 +80,15 @@ rustPlatform.buildRustPackage {
     libxkbcommon
   ];
 
+  # Required but not explicitly requested libraries
+  runtimeDependencies = [
+    xorg.libX11
+    xorg.libXi
+    xorg.libXcursor
+    xorg.libXrandr
+    vulkan-loader
+  ];
+
   buildNoDefaultFeatures = true;
   buildFeatures = [ "default-publish" ];
 
@@ -97,17 +111,6 @@ rustPlatform.buildRustPackage {
   # Some tests require internet access
   doCheck = false;
 
-  postFixup = ''
-    # Add required but not explicitly requested libraries
-    patchelf --add-rpath '${lib.makeLibraryPath [
-      xorg.libX11
-      xorg.libXi
-      xorg.libXcursor
-      xorg.libXrandr
-      vulkan-loader
-    ]}' "$out/bin/veloren-voxygen"
-  '';
-
   postInstall = ''
     install -Dm644 assets/voxygen/net.veloren.veloren.desktop -t "$out/share/applications"
     install -Dm644 assets/voxygen/net.veloren.veloren.png "$out/share/pixmaps"
```

tomodachi94 reviewed (changes requested):
Some minor concerns. Additionally:
* It looks like [0.16.0](https://gitlab.com/veloren/veloren/-/releases/v0.16.0) has been released. Can you update the package to this version?
* Could you reformat the package with `nixfmt-rfc-style`?
* Could you rebase this onto the latest master, since it's been a while since this was opened?

rnhmjoj commented:
Rebased and updated to 0.16.0. Every update is a real pain because rust is the apotheosis of "move fast and break things", but this was somewhat faster... probably because I know the drill now.

rnhmjoj commented:
... and formatted with nixfmt.

Anyway, the problem of the assets remains. It would be good to not dump on hydra/binary cache a 400 MB derivation of static files on every update. I haven't figured out how to discard it while at the time depending on it.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-exclude-derivation-in-a-package-closure-from-the-binary-cache/50144/1

rnhmjoj commented:
I had managed to:
1. remove the assets from the source
2. fetching them with another FOD with `lib.dontDistribute`
3. using `buildEnv` to join the game and the assets without creating a runtime dependency
4. patching the game to find the assets directory by going up relative to argv[0]

but then I found out it needs access to the assets while compiling some subpackage. grrr...
This makes the main derivation depend on the assets at build time, so hydra still needs to fetch them making all of the above pointless.

So, fine, I give up on not wasting disk space.

rnhmjoj commented:
@tomodachi94 are you interested in maintaining this? Updating this package is not exactly trivial, so some help is more than welcome.

tomodachi94 commented:
> @tomodachi94 are you interested in maintaining this? Updating this package is not exactly trivial, so some help is more than welcome.

Sure. I don't have a personal stake in Veloren (I haven't played it yet), but I'm planning on playing it at some point.

tomodachi94 reviewed (approved):
Lgtm. I'm not well-versed in Rust, but seems okay.

rnhmjoj commented:
Can I add you to the maintainers list?

tomodachi94 commented:
> Can I add you to the maintainers list?

Yes, please do :)
